src/components/ClapButton.jsx

```diff
@@ -0,0 +1,145 @@
+/* eslint-disable no-undef */
+import React, { Component } from 'react';
+import PropType from 'prop-types';
+import { connect } from 'react-redux';
+import { createStructuredSelector } from 'reselect';
+import { debounce } from 'lodash';
+import { withRouter } from 'react-router-dom';
+import clapIcon from '../images/clap.svg';
+import * as articleSelector from '../selectors/singleArticleSelector';
+import { getAuthId, getIsAuthenticated } from '../selectors/authSelector';
+import { getToken } from '../selectors/profileSelector';
+import { clapArticleRequest, fetchUserClaps } from '../actions/clapsAction';
+import { getUserClaps } from '../selectors/clapsSelector';
+import { CLAPS_LIMIT, CLAPS_REQUEST_TIMER } from '../utils/contants';
+
+/**
+ * A class representing clap button component
+ *
+ * @class ClapButton
+ * @extends {Component}
+ */
+export class ClapButton extends Component {
+  /**
+   * Handles clap request (debounced)
+   *
+   * @returns {void}
+   * @memberof ClapButton
+   */
+  makeClapRequest = debounce(() => {
+    const { clapArticle, articleSlug: slug, token } = this.props;
+    const { clapsCount: claps } = this.state;
+    this.setState({ clapsCount: 0 });
+    return clapArticle({ slug, claps, token });
+  }, CLAPS_REQUEST_TIMER);
+
+  static propTypes = {
+    articleSlug: PropType.string.isRequired,
+    articleAuthorId: PropType.number.isRequired,
+    articleClaps: PropType.number.isRequired,
+    clapArticle: PropType.func.isRequired,
+    isLoggedIn: PropType.bool.isRequired,
+    userId: PropType.number,
+    history: PropType.object.isRequired,
+    token: PropType.string.isRequired,
+    userClaps: PropType.number.isRequired,
+    loadUserClaps: PropType.func.isRequired,
+  };
+
+  static defaultProps = {
+    userId: null,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      clapsLeft: CLAPS_LIMIT - props.userClaps,
+      articleClaps: props.articleClaps,
+      clapsCount: 0, // record clap button clicks
+    };
+  }
+
+  componentDidMount = async () => {
+    const {
+      articleSlug: slug,
+      isLoggedIn,
+      userId,
+      loadUserClaps
+    } = this.props;
+    if (isLoggedIn) {
+      loadUserClaps({ slug, userId });
+    }
+  }
+
+  /**
+   * Makes an article clap
+   *
+   * @returns {void}
+   * @memberof ClapButton
+   */
+  handleClapArticle = () => {
+    const {
+      isLoggedIn,
+      userId,
+      articleAuthorId,
+      history,
+    } = this.props;
+
+    if (!isLoggedIn) { // redirect guest users to login
+      history.push('/login');
+      return false;
+    }
+    // Deny article's author
+    if (articleAuthorId === userId) return false;
```

Comment:
**Nitpick**


- Right now if an author claps on his/her own article, nothing happens. Do you think we can add visual feedback to inform the author that they can not clap on their own article?

Reply:
That's a good idea @valenstical. We'll popup a toaster informing the author that they cannot clap their article.